scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
+        return membership.status == MembershipStatus.Owner
+    except ObjectDoesNotExist as dne:
+        # if there is no record the user is not in the group and cannot be an owner
+        return False
+
+
+class KafkaTopic(models.Model):
+    owning_group = models.ForeignKey(
+        Group,
+        on_delete=models.CASCADE,
+    )
+    name = models.CharField(
+        max_length=249, # see https://github.com/apache/kafka/commit/ad3dfc6ab25c3f80d2425e24e72ae732b850dc60
+        editable=False,
+    )
+    publicly_readable = models.BooleanField(
+        default = False,
+    )
+
+
+def validate_topic_name(name: str) -> bool:
+    # https://github.com/apache/kafka/blob/bc55f85237cb46e73c6774298cf308060a4a739c/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L30
+    valid = re.compile("^[a-zA-Z0-9._-]{1,249}$")
+    return re.match(valid, name)
+
+
+class KafkaOperation(enum.Enum):
+    All = 1
+    Read = 2
+    Write = 3
+    Create = 4
+    Delete = 5
+    Alter = 6
+    Describe = 7
+    ClusterAction = 8
+    DescribeConfigs = 9
+    AlterConfigs = 10
+    IdempotentWrite = 11
+
+
+class GroupKafkaPermission(models.Model):
+    principal = models.ForeignKey(
+        Group,
+        on_delete=models.CASCADE,
+    )
+    topic = models.ForeignKey(
+        KafkaTopic,
+        on_delete=models.CASCADE,
+    )
+    operation = enum.EnumField(
+        KafkaOperation,
+        editable=False,
+    )
+
+
+def equivalent_permission(p1: GroupKafkaPermission, 
+                          p2: GroupKafkaPermission) -> bool:
+    return p1.principal==p2.principal \
+           and p1.topic==p2.topic \
+           and p1.operation==p2.operation
+
+
+def addKafkaPermissionForGroup(group_id: str, topic: KafkaTopic, operation: KafkaOperation):
+    group = Group.objects.get(id=group_id)
+    new_record = GroupKafkaPermission(principal=group, topic=topic, operation=operation)
+    
+    # look up all permissions for this group/topic combination to figure out if the new record is 
+    # redundant or some old ones need to be replaced
+    existing = GroupKafkaPermission.objects.filter(principal=group_id, topic=topic.id)
+    
+    if len(existing)>0:
+        if any(equivalent_permission(p, new_record) for p in existing):
+            # the exact permission we're trying to create already exists, 
+            # so we can do nothing and declare success
+            return
+        if len(existing)==1 and existing[0].operation==KafkaOperation.All:
+            # the existing permission is broader than the one being added,  
+            # so we do not need to actually add it
+            return
+        if operation==KafkaOperation.All:
+            # the new permission will supercede all existing permissions, so we remove them
+            for old in existing:
+                old.delete()
+    
+    # any possible redundancy being resolved, we can now create the actual record
+    new_record.save()
```

Comment:
As mentioned previously, I would like to make this properly concurrency-safe, but I have no previous experience doing that in the context of SQL databases/django, so I'm not sure if we want to hold up these features while I learn.

Reply:
It might not be too bad, actually. Django's DB transactions API looks a lot better than when I last used it 9 years ago! https://docs.djangoproject.com/en/3.1/topics/db/transactions/

I sat down to try my hand at it. How about this:
```python
from django.db import transaction
def addKafkaPermissionForGroup(group_id: str, topic: KafkaTopic, operation: KafkaOperation):
    existing = GroupKafkaPermission.objects.filter(principal=group, topic=topic.id)

    # Don't do anything if the permissions already exist.
    if existing.filter(operation=operation).exists():
        return
   
    # Don't do anything if the group already has "All" permission.
    if existing.filter(operation=KafkaOperation.All).exists():
        return

    # Okay, we're definitely adding a record.
    group = Group.objects.get(id=group_id)
    new_record = GroupKafkaPermission(principal=group, topic=topic, operation=operation)

    with transaction.atomic():
        new_record.save()
        # If we wrote an "All" permission, then clean up any other permissions since 
        # they're redundant.
        if new_record.operation == KafkaOperation.All:
             existing.exclude(operation=KafkaOperation.All).delete()
```

Reply:
I've wrapped each of the cases of writing to more than one record with an atomic transaction block, which should improve matters somewhat. Unfortunately, this still leaves us unprotected against concurrent changes from independent requests, which may be something to come back to later.